Customize format for I32/U32 (issue #1920)

krig opened the pull request:
Fixes formatting of I32.min_value().

TODO: the u32 and u64 funs in `_format_int.pony` could probably be largely re-unified.
    
Fixes: #1920

SeanTAllen commented:
There was a TravisCI failure that was unrelated to code, it was during dependency install. Restarted that one job.

SeanTAllen commented on packages/format/_format_int.pony:
```diff
@@ -65,6 +65,40 @@ primitive _FormatInt
 
+  fun u32(x: U32, neg: Bool, fmt: FormatInt, prefix: PrefixNumber,
+    prec: USize, width: USize, align: Align, fill: U32
+  ): String iso^ =>
```
see the style guide that is in progress for formatting on this.

krig replied:
Updated the PR. Something like this?

SeanTAllen commented:
That would fit the forthcoming style guide @krig. Yes.

krig commented:
Alright, great. Fixed one small formatting mistake in the previous update.

SeanTAllen commented on packages/format/format.pony:
```diff
@@ -96,8 +96,10 @@ primitive Format
     elseif x is I128 then
       _FormatInt.u128(abs.u128(), neg, fmt, prefix, prec, width, align, fill)
-    else
+    elseif (x is U64) or (x is I64) then
```
there's a bug that predates your code @krig. These should be using a match expression or an `iftype`. The latter is preferred.

The `is` doesn't work as it compares identity rather than checking the type.

So `x is I128` will only be triggered for the value 0. otherwise the else will always be used.

krig commented:
@SeanTAllen I fixed the if/iftype bug, at least I think so. At the same time I thought to test if formatting min_value works for I8/I16 as well, since I suspected it didn't. And it didn't. So I implemented those as well. Ideas for how to reduce the duplication across the various int functions are welcome, I'm still very much a pony newbie :)

krig commented:
Hm. It also doesn't work properly for `ISize` or `ILong`. Not sure how to solve those.

krig commented:
@jemc Ah, perfect! I'll update it.

jemc commented on packages/format/format.pony:
```diff
@@ -92,12 +92,26 @@ primitive Format
+      _FormatInt.u16(abs.u16(), neg, fmt, prefix, prec, width, align, fill)
+    elseif A <: (U8 | I8) then
+      _FormatInt.u8(abs.u8(), neg, fmt, prefix, prec, width, align, fill)
+    elseif A <: (USize | ISize | ULong | ILong) then
```
Sorry, this isn't quite right - the `*Size` types have different bitwidth semantics from the `*Long` types.  Please see the definition of `ISize.bitwidth()`.

krig replied:
Ah, I see, thanks.

jemc commented:
Sorry, looks like there's one more issue in the tests, where you're testing `ISize` but using the `ifdef` rule for `ILong`.

It's causing failures on the windows CI builds.

krig commented:
Good thing I added the test.. alright, this should do it.

jemc commented:
Great, thanks for bringing this one home!

krig commented:
Cool, thank you both!

SeanTAllen commented:
Awesome to see this merged! Thanks @krig. You rock.
